src/compiler/checker.ts

```diff
@@ -27534,7 +27534,7 @@ namespace ts {
                 const inferenceContext = getInferenceContext(node);
                 // If no inferences have been made, nothing is gained from instantiating as type parameters
                 // would just be replaced with their defaults similar to the apparent type.
-                if (inferenceContext && contextFlags! & ContextFlags.Signature && some(inferenceContext.inferences, hasInferenceCandidates)) {
+                if (inferenceContext && (contextFlags! & ContextFlags.Signature && some(inferenceContext.inferences, hasInferenceCandidates) || maybeTypeOfKind(contextualType, TypeFlags.Conditional))) {

```

Comment:
I imagine that `IndexedAccess` and some other types that try to "route" a type should also be included here, in the fix.